File: ios/Classes/api/ViewCanvasApi.swift
//
//  ViewCanvasApi.swift
//  jc
//
//  Created by Михаил on 13.04.2023.
//

import Foundation
import JCSDKOC

protocol ViewCanvasApi {
    func startVideo(mode: JCMediaDeviceRender) -> JCMediaDeviceVideoCanvas?
}

class SelfViewCanvasApi : ViewCanvasApi {
    private var oneToOne: SelfViewCanvasApiOneToOne = SelfViewCanvasApiOneToOne()
    private var group: SelfViewCanvasApiGroup = SelfViewCanvasApiGroup()

    func startVideo(mode: JCMediaDeviceRender) -> JCMediaDeviceVideoCanvas? {
        switch (JCCallUtils.getCallType()) {
        case .oneToOne: return oneToOne.startVideo(mode: mode)
        case .group: return group.startVideo(mode: mode)
        case .none:
            return nil
        }
    }
}

class SelfViewCanvasApiOneToOne : ViewCanvasApi {
    func startVideo(mode: JCMediaDeviceRender) -> JCMediaDeviceVideoCanvas? {
        return JCRoom.shared.call.getActiveCallItem()?.startSelfVideo(mode)
    }
}

class SelfViewCanvasApiGroup : ViewCanvasApi {
    func startVideo(mode: JCMediaDeviceRender) -> JCMediaDeviceVideoCanvas? {
        let selfPart = JCRoomUtils.selfParticipant!
        let canvas = selfPart.startVideo(
            mode,
            pictureSize: .large
        )
        JCRoom.shared.mediaChannel.requestVideo(selfPart, pictureSize: .large)
        return canvas
    }
}

class CompanionViewCanvasApi : ViewCanvasApi {
    private var oneToOne: CompanionCanvasApiOneToOne = CompanionCanvasApiOneToOne()
    private var group: CompanionViewCanvasApiGroup = CompanionViewCanvasApiGroup()

    func startVideo(mode: JCMediaDeviceRender) -> JCMediaDeviceVideoCanvas? {
        switch (JCCallUtils.getCallType()) {
        case .oneToOne: return oneToOne.startVideo(mode: mode)
        case .group: return group.startVideo(mode: mode)
        case .none:
            return nil
        }
    }
}

class CompanionCanvasApiOneToOne {
    func startVideo(mode: JCMediaDeviceRender) -> JCMediaDeviceVideoCanvas? {
        return JCRoom.shared.call.getActiveCallItem()?.startOtherVideo(mode)
    }
}

class CompanionViewCanvasApiGroup {
    func startVideo(mode: JCMediaDeviceRender) -> JCMediaDeviceVideoCanvas? {
        return JCCallUtils.otherParticipant?.startVideo(
            mode,
            pictureSize: .large
        )
    }
}

File: android/src/main/kotlin/com/michaellazebny/jyphoon/jc/api/ViewCanvasApi.kt
package com.michaellazebny.jyphoon.jc.api

import android.view.View
import com.juphoon.cloud.JCMediaChannel
import com.michaellazebny.jyphoon.jc.CallType
import com.michaellazebny.jyphoon.jc.jcWrapper.JCManager
import com.michaellazebny.jyphoon.jc.utils.JCCallUtils

sealed class ViewCanvasApi {
    abstract fun startVideo(mode: Int) : View?

    class SelfViewCanvasApi : ViewCanvasApi() {
        private val oneToOne: SelfViewCanvasApiOneToOne = SelfViewCanvasApiOneToOne()
        private val group: SelfViewCanvasApiGroup = SelfViewCanvasApiGroup()

        override fun startVideo(mode: Int): View? {
            return when (JCCallUtils.getCallType()) {
                CallType.ONETOONE -> oneToOne.startVideo(mode)
                CallType.GROUP -> group.startVideo(mode)
                else -> null
            }
        }

        class SelfViewCanvasApiOneToOne {
            fun startVideo(mode: Int): View? {
                return JCManager.getInstance().call.activeCallItem.startSelfVideo(mode)?.videoView
            }
        }

        class SelfViewCanvasApiGroup  {
            fun startVideo(mode: Int): View? {
                return JCManager.getInstance().mediaChannel.selfParticipant?.startVideo(
                    mode,
                    JCMediaChannel.PICTURESIZE_MIN
                )?.videoView
            }
        }
    }

    class CompanionViewCanvasApi : ViewCanvasApi() {
        private val oneToOne: CompanionCanvasApiOneToOne = CompanionCanvasApiOneToOne()
        private val group: CompanionViewCanvasApiGroup = CompanionViewCanvasApiGroup()

        override fun startVideo(mode: Int): View? {
            return when (JCCallUtils.getCallType()) {
                CallType.ONETOONE -> oneToOne.startVideo(mode)
                CallType.GROUP -> group.startVideo(mode)
                else -> null
            }
        }

        class CompanionCanvasApiOneToOne {
            fun startVideo(mode: Int): View? {
                return JCManager.getInstance().call.activeCallItem.startOtherVideo(mode)?.videoView
            }
        }

        class CompanionViewCanvasApiGroup {
            fun startVideo(mode: Int): View? {
                return JCCallUtils.otherParticipant?.startVideo(
                    mode,
                    JCMediaChannel.PICTURESIZE_MIN
                )?.videoView
            }
        }
    }
}
